doc/cprover-manual/function-contracts.md

```diff
@@ -0,0 +1,502 @@
+fully specify these operations. Therefore, CBMC also provide a series of
+built-in constructus to be used with functions contracts (e.g., _historic
+variables_, _quantifiers_, and _memory predicates_).
+
+
+## Contract Clauses
+
+Let's write function contracts, based on the English specification in the
+comments, for the following example in order to properly demonstrate the
+semantics of the basic contract clauses in CBMC: `__CPROVER_requires`,
+`__CPROVER_ensures`, and `__CPROVER_assigns`.
+
+```c
+#include <stddef.h>
+#include <stdint.h>
+
+#define SUCCESS 0
+#define FAILURE -1
+
+/**
+ * If a + b overflows, returns FAILURE; otherwise sums
+ * a + b, returns the result in *out, and returns SUCCESS.
+ */
+int add_overflow(uint32_t a, uint32_t b, uint32_t* out)
+{
+    /* ... */
+}
+
+```
+
+The _requires_ clause specify the preconditions of a function, i.e., the minimal
+requirements to properly execute an operation. Developers may see the _requires_
+clause as all the caller's obligations when invoking this fuction. In our
+example, the developer may require from the caller to properly allocate all
+arguments, thus, pointers must be valid. For now, let's require that `out` is
+different than `NULL`, but we can improve this example later with additional
+_memory primitives_.
+
+```c
+int add_overflow(uint32_t a, uint32_t b, uint32_t* out)
+__CPROVER_requires(out != NULL)
+{
+    /* ... */
+}
+```
+
+The _ensures_ clause specify the postconditions of a function, i.e., all the
+properties over arguments or global variables that a function guarantees at the
+an of the operation.  In our example, developers can use the built-in construct
+`__CPROVER_return_value`, which represents the return value of a function. As
+postconditions, one may list a possible range for the return values (in this
+case, either `SUCCESS` or `FAILURE`) as well as describe the main property of
+this function, i.e., if the function returns `SUCCESS`, then `*out` stores the
+result of `a + b`. We write the later as an implication, since _requires_ and
+_ensures_ clauses only accept Boolean statements. Also, notice that one may call
+multiple instances of `__CPROVER_requires ` and `__CPROVER_ensures `, which will
+be later combined together in a single conjunction.
+
+```c
+int add_overflow(uint32_t a, uint32_t b, uint32_t* out)
+__CPROVER_requires(out != NULL)
+__CPROVER_ensures(__CPROVER_return_value == SUCCESS || __CPROVER_return_value == FAILURE)
+__CPROVER_ensures((__CPROVER_return_value == SUCCESS) ==> (*out == (a + b)))
+{
+    /* ... */
+}
+```
+**Important.** Developers may call functions inside _requires_ and _ensures_
+clauses to better write larger specifications (e.g., predicates). However, at
+this point CBMC does not enforce such functions to be **pure** (i.e., without
+side effects). This will be supported in future versions.
+
+Finally, the _assigns_ clause allows developers to specify a list of memory
```

Comment:
Comment: Should we add that we only do assigns clauses for function arguments?

Reply:
We also accept global variables. Fixed it.